treeset/src/main/java/ru/hse/mnmalysheva/treeset/TreapSet.java

```diff
@@ -0,0 +1,433 @@
+            return node.data;
+        }
+        return nextNode(node).data;
+    }
+
+    private NodePair split(Node root, E element) {
+        if (root == nullNode) {
+            return new NodePair(nullNode, nullNode);
+        }
+        if (compare(element, root.data) <= 0) {
+            NodePair result = split(root.left, element);
+            root.setLeft(result.right);
+            result.setRight(root);
+            return result;
+        }
+        NodePair result = split(root.right, element);
+        root.setRight(result.left);
+        result.setLeft(root);
+        return result;
+    }
+
+    private Node merge(Node left, Node right) {
+        if (left == nullNode || right == nullNode) {
+            return left != nullNode ? left : right;
+        }
+        if (left.priority > right.priority) {
+            left.setRight(merge(left.right, right));
+            return left;
+        }
+        right.setLeft(merge(left, right.left));
+        return right;
+    }
+
+    private Node descentTo(Node root, Object element) {
+        if (compare(root.data, element) < 0) {
+            return root.right == nullNode ? root : descentTo(root.right, element);
+        }
+        if (compare(root.data, element) > 0) {
+            return root.left == nullNode ? root : descentTo(root.left, element);
+        }
+        return root;
+    }
+
+    private Node descentTo(Object element) {
+        return root == nullNode ? nullNode : descentTo(root, element);
+    }
+
+    private Node firstNode(Node root) {
+        var node = root;
+        while (node.left != nullNode) {
+            node = node.left;
+        }
+        return node;
+    }
+
+    private Node lastNode(Node root) {
+        var node = root;
+        while (node.right != nullNode) {
+            node = node.right;
+        }
+        return node;
+    }
+
+    private Node previousNode(Node node) {
+        if (node.left != nullNode) {
+            return lastNode(node.left);
+        }
+        while (node.parent != nullNode && node != node.parent.right) {
+            node = node.parent;
+        }
+        return node.parent;
+    }
+
+    private Node nextNode(Node node) {
+        if (node.right != nullNode) {
+            return firstNode(node.right);
+        }
+        while (node.parent != nullNode && node != node.parent.left) {
+            node = node.parent;
+        }
+        return node.parent;
+    }
+
+    @SuppressWarnings("unchecked")
+    private int compare(Object first, Object second) {
+        if (comparator == null) {
+            return ((Comparable) first).compareTo(second);
+        }
+        return comparator.compare((E) first, (E) second);
+    }
+
+    private class Node {
+        private Node left;
+        private Node right;
+        private Node parent;
+        private E data;
+        private int priority;
+        private int subtreeSize;
+
+        // null node constructor
```

Comment:
Мне кажется, было бы странно, если бы ровно к одному приватному методу был джавадок, а к другим нет, но при этом писать джавадоки ко всем приватным методам не хочется.
Вообще, не очень понятно, что делать, если хочется прокомментировать неочевидное место в коде. Здесь так совпало, что комментарий ко всему методу, но может быть и просто к какой-то строчке внутри него. В качестве возможного решения вижу только "не писать неочевидный код", и оно мне кажется разумным, но не уверена, что это всегда возможно. Но в данном случае возможно, пофикшу.

Reply:
Если есть строчка, в которой хочется написать коммент, сделайте метод, название которого будет отражать то, что он делает. Типа:
не так: a = a + 1 // this is because we are currently on Windows
а так: a = adjustToWindows(a);
Это все разумеется близко к области холивора, но, мне кажется, с методом все же лучше. Ну а к одному приватному методу написать джавадок, мне не кажется страшным. Это даже нужно, если приватный метод нетривальный и делает что-то хитрое.

Reply:
Спасибо